### ISSUE TITLE
Possible bug in string operation substr

### ISSUE THREAD

**Issue by OrenGitHub** [Z3Prover/z3#1415]
Hi,
I'm running the following query:

    (declare-const si String)
    (declare-const sj String)
    
    (assert (=  2 (str.indexof si "#"    0)))
    (assert (= -1 (str.indexof si "\x00" 0)))
    (assert (= (str.substr si 0 2) (str.substr sj 0 2)))
    (assert (= (str.substr si 5 5) (str.substr sj 5 5)))
    
    (check-sat)
    (get-value (si sj))

But I get **wrong assignments** for the solution:

    sat
    ((si " \x04#\x08")
     (sj ""))

When I change the second sub-string constraint to **start at 4** like this:

    (declare-const si String)
    (declare-const sj String)
    
    (assert (=  2 (str.indexof si "#"    0)))
    (assert (= -1 (str.indexof si "\x00" 0)))
    (assert (= (str.substr si 0 2) (str.substr sj 0 2)))
    (assert (= (str.substr si 4 5) (str.substr sj 4 5)))
    
    (check-sat)
    (get-value (si sj))

Then I get a solution that  makes more sense
(though **I would expect si and sj to be with lengths 9 at least**, no?)

    sat
    ((si "\x08\x04#\x02�")
     (sj "\x08\x04\x00\x00�"))

**Comment by pjljvandelaar** [comment]
See http://cvc4.cs.stanford.edu/wiki/Strings.
Substring doesn't enforce a length of 9.
Only that when the string has characters in the range, the characters have to adhere to the assertion.

**Comment by OrenGitHub** [comment]
@pjljvandelaar  thanks, what about the first query?
That assignment there is a bug, right?

**Comment by pjljvandelaar** [comment]
I agree with you and think it is a bug. Let's hope it will be fixed soon.
Does cvc4 yield the right answer?

**Comment by pjljvandelaar** [comment]
And with z3str3 what is the answer?

**Comment by iwob** [comment]
When I tried the query on HEAD, I got the following model:
```
((si "\x00\x08#@")
 (sj "\x00"))
```
which is wrong because there should be no \x00.

Model by Z3 with smt.string_solver=z3str3:
```
((si "ah##r")
 (sj "ah"))
```

Model by CVC4 (nightly build):
```
((si "AB#") (sj "AB"))
```

**Comment by NikolajBjorner** [comment]
it is fixed now with 45b6e0998aef1971bcda83ed3e8e08639efadb42
